service/history/task/redispatcher.go

```diff
@@ -146,21 +134,43 @@ func (r *redispatcherImpl) AddTask(task Task) {
 	r.timerGate.Update(t)
 }
 
-// TODO: review this method, it doesn't seem to redispatch the tasks immediately
+// TODO: this method is removed from Temporal's codebase in this PR: https://github.com/temporalio/temporal/pull/3038
+// Let's review if this method is needed in our codebase later.
 func (r *redispatcherImpl) Redispatch(targetSize int) {
-	doneCh := make(chan struct{})
-	ntf := redispatchNotification{
-		targetSize: targetSize,
-		doneCh:     doneCh,
+	r.Lock()
+	defer r.Unlock()
+
+	queueSize := r.sizeLocked()
+	r.metricsScope.RecordTimer(metrics.TaskRedispatchQueuePendingTasksTimer, time.Duration(queueSize))
+
+	now := r.timeSource.Now()
+	totalRedispatched := 0
+	var failToSubmit []redispatchTask
+	for !r.pq.IsEmpty() && totalRedispatched < targetSize {
+		item, _ := r.pq.Peek() // error is impossible because we've checked that the queue is not empty
+		if item.redispatchTime.After(now) {
+			break
+		}
+		submitted, err := r.taskProcessor.TrySubmit(item.task)
```

Comment:
so, I guess the plan for those tasks that couldn't be submitted is to rely on the redispatch timer? 

The core semantic change  - as I understand it - is that you're making a function that may fail to redispatch the tasks via the async loop, but then puts it in a in-memory queue for redispatch next tick? 

that seems reasonble, but just making sure I'm following.